File: kattis/gigcombinatorics.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
#define MAX 1000010

int N, nums[MAX];
ll memo[2][MAX];
const ll MOD = 1e9 + 7;

inline ll mult(ll a, ll b) { return (a * b) % MOD; }
inline ll add(ll a, ll b) { return (a + b) % MOD; }

ll dp(bool first, int n) {
  if (n == N) return 0;
  if (~memo[first][n]) return memo[first][n];

  ll total = dp(false, n + 1);

  if (nums[n] == 2 && !first)
    total = mult(total, 2);
  else if (nums[n] == 3)
    total = add(total, 1);

  return memo[first][n] = total;
}

int main() {
  while (scanf("%d", &N) == 1) {
    memset(memo, -1, sizeof memo);
    for (int i = 0; i < N; i++) scanf("%d", &nums[i]);

    ll total = 0;
    ll ones = 0;
    for (int i = 0; i < N; i++)
      if (nums[i] == 1)
        ones++;
      else if (nums[i] == 2)
        total = add(total, mult(ones, dp(true, i)));

    printf("%lld\n", total);
  }
}

File: kattis/gigcombinatorics.py
MOD = int(1e9 + 7)
nums = []
memo = []
N = None


def dp(n):
    if n == N:
        return 0
    if memo[n] != -1:
        return memo[n]

    total = dp(n + 1)

    if nums[n] == 2:
        total *= 2
    elif nums[n] == 3:
        total += 1

    memo[n] = total % MOD

    return memo[n]


while(True):
    try:
        line = input()
    except EOFError:
        break

    N = int(line)

    nums = list(map(int, input().split()))

    memo = [-1] * N

    total = 0
    ones = 0

    for i in reversed(range(N)):
        dp(i)

    for i in range(N):
        if nums[i] == 1:
            ones += 1
        elif nums[i] == 2:
            total += ones * dp(i + 1)
            total %= MOD

    print(total % MOD)
